### PR TITLE
Add runtime profiler for collecting runtime stats in parts of the code

### PR DESCRIPTION
**Pull request by NathanW2**
This PR is mainly a proof of concept at the moment (so ignore the code for now)  but keen to get feedback as I think it would be useful going forward

PR adds a profiling class which is mainly used to collect timings around parts of the code.  It will be exposed via the API so other parts of the code can use it and we can collect code timings in a single place for all parts of the application using the same method of timing.

Keeping the API simple for now but could add groups later on.

Basic usage at the moment is

```
profiler->start("Some message");
// Do something
profiler->end();
```

I have already add it to `QgisApp` to profile the start up times there.  Having timings around the main parts of the code is handy to look for slow downs after new features and to see if we can trim it down.

Example of log output however this could be shown in the user interface someplace if we need it

```
..\src\app\qgisapp.cpp(1029) : (QgisApp::QgisApp) [0ms] PROFILE TIMES
..\src\app\qgisapp.cpp(1030) : (QgisApp::QgisApp) [0ms] PROFILE TIMES TOTAL - 12.811 
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - SETTING UP UI - 0.208
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CHECKING DATABASE - 0.037
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - INITIALIZING AUTHENTICATION - 1.084
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATING THEME FOLDER - 0.003
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - BUILDING STYLE SHEET - 0.009
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATING MAP CANVAS - 0.017
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - WELCOME PAGE - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [1ms]  - MESSAGE BAR - 0
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - USER INPUT DOCK - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - LAYER TREE - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - UNDO DOCK - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - ADVANCED DIGITIZE PANEL - 0.006
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - STATS DOCK - 0.014
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - BOOKMARKS WIDGET - 0.019
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - SNAPPING UTILS - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATE ACTIONS - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATE ACTION GROUP - 0
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATE MENUS - 0.002
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - TOOLBARS - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - STATUS BAR - 0.01
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [1ms]  - CREATE CANVAS TOOLS - 0.016
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - INIT LAYER TREE VIEW - 0.018
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATE OVERVIEW - 0.002
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATE MAP TIPS - 0
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - CREATE DECORATIONS - 0.002
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - READ SETTINGS - 0.319
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - UPDATE RECENT PROJECT PATHS - 0
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - UPDATE PROJECT FROM TEMPLATES - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - LEGEND LAYER SELECTION CHANGED - 0
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - PLUGIN MANAGER - 0.063
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - MAP STYLE DOCK - 0.005
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - SNAPPING DIALOG - 0.005
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - INITPLUGININSTALLER - 10.203
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - RESTORE WINDOW STATE - 0.005
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - UPDATE CUSTOMIZIATION ON MAIN WINDOW - 0.001
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [1ms]  - SHOW MAIN WINDOW - 0.735
..\src\app\qgisapp.cpp(1036) : (QgisApp::QgisApp) [0ms]  - NEW PROJECT - 0.02
```

### CODE REVIEW

**Comment by nirvn** [comment]
10.2s for initialisation of plugin installer? That's probably worth investigating.

**Comment by rouault** [comment]
Have you considered situations where calls to the profile could be nested, like :
```
profiler->start("my_complicated_processing");
my_complicated_processing();
profile->stop();

my_complicated_processing()
{
  profiler->start("step1");
  ....
  profiler->stop();
  profiler->start("step2");
  ....
  profiler->stop();
}
```

(Oh I see now your sentence about groups : I guess that must match the above idea)

Also thinking that making a C++ object to be sure to have correct pairing could be interesting. At least if you want to have profiling for a whole method and may have seveal return points.

```
my_function()
{
  ProfilerStep profilerStep("step_name");
  ...
}
```
For several profiling steps inside the same method, it is less attractive as it requires to add blocks.

I don't really have use cases in mind, just throwing out potential ideas.

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -978,6 +1026,16 @@ QgisApp::QgisApp( QSplashScreen *splash, bool restorePlugins, bool skipVersionCh
 #ifdef ANDROID
   toggleFullScreen();
 #endif
+  QgsDebugMsg("PROFILE TIMES");
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: I'd move this to the profiler class. Then add a nice ascii-art style bar chart to make it obvious where the bottlenecks are, eg:

```
PROFILE TIMES TOTAL - 12.811 
00o........ - SETTING UP UI - 0.208
o.......... - CHECKING DATABASE - 0.037
0000oo..... - CREATING THEME FOLDER - 0.003
00000000000 - some slow thing - 1111.01
etc
```

**Comment by NathanW2** [comment]
@rouault I was tempted to solve it the same was a QSettings does. Something like:

```
start("group\details")
```

so for the example

```
profiler->start("processing\my_complicated_processing");
my_complicated_processing();
profile->stop();

my_complicated_processing()
{
  profiler->start("processing\step1");
  ....
  profiler->stop();
  profiler->start("processing\step2");
  ....
  profiler->stop();
}
```

and just split on "\" when counting etc if you need to collect info on the group for total, etc.  Of course there is a lot of other ways one can do this, mainly a trade off between easy API use and more control.

**Comment by NathanW2** [comment]
@nirvn yeah plugin loading is super heavy. I also have timings in the Python code to profile the load times of plugins.  If we use this method I can move there timings up to use this class so that we can see the plugin break down.

Come to think of it, it's a clear case for the nested groups that @rouault talked about.  I might test that theory and update the PR to see how people fell about it.

File: src/app/qgisapp.cpp (hunk 1 above)
Comment (NathanW2) [inline src/app/qgisapp.cpp]: Good idea.  I was also going to just have a few other "reporting" methods, e.g `max`, `top10hot`, `asHtml`, etc.

Maybe it also makes sense to store the data for the runs in a sqlite db with a data time stamp for the profile instance.

**Comment by NathanW2** [comment]
@rouault how do you feel about how QSettings does it? http://doc.qt.io/qt-4.8/qsettings.html#beginGroup

```
profiler.beginGroup("init");
profiler.endGroup()
```

It could just push to a stack of groups so it can be nested and then just joined into a single string at the end: `init/python/blah`

**Comment by rouault** [comment]
> how do you feel about how QSettings does it? http://doc.qt.io/qt-4.8/qsettings.html#beginGroup
LGTM

**Comment by m-kuhn** [comment]
Seems to be merged already

**Comment by NathanW2** [comment]
Thanks was just coming back to close but got distracted :)

**Pull request closed without merging**